client/src/main/java/com/aliyun/emr/rss/client/read/RemotePartitionReader.java

```diff
@@ -0,0 +1,136 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.aliyun.emr.rss.client.read;
+
+import java.io.IOException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.netty.buffer.ByteBuf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.aliyun.emr.rss.common.RssConf;
+import com.aliyun.emr.rss.common.network.buffer.ManagedBuffer;
+import com.aliyun.emr.rss.common.network.buffer.NettyManagedBuffer;
+import com.aliyun.emr.rss.common.network.client.ChunkReceivedCallback;
+import com.aliyun.emr.rss.common.network.client.TransportClientFactory;
+import com.aliyun.emr.rss.common.protocol.PartitionLocation;
+
+public class RemotePartitionReader implements PartitionReader {
+  private final Logger logger = LoggerFactory.getLogger(RemotePartitionReader.class);
+  private final RetryingChunkClient client;
+  private final int numChunks;
+
+  private int returnedChunks;
+  private int chunkIndex;
+
+  private final LinkedBlockingQueue<ByteBuf> results;
+  private final ChunkReceivedCallback callback;
+
+  private final AtomicReference<IOException> exception = new AtomicReference<>();
+  private final int maxInFlight;
+  private boolean closed = false;
+
+  RemotePartitionReader(
+    RssConf conf,
+    String shuffleKey,
+    PartitionLocation location,
+    TransportClientFactory clientFactory,
+    int startMapIndex,
+    int endMapIndex) throws IOException {
+    maxInFlight = RssConf.fetchChunkMaxReqsInFlight(conf);
+    results = new LinkedBlockingQueue<>();
+    callback = new ChunkReceivedCallback() {
+      @Override
+      public void onSuccess(int chunkIndex, ManagedBuffer buffer) {
+        // only add the buffer to results queue if this reader is not closed.
+        synchronized (this) {
+          ByteBuf buf = ((NettyManagedBuffer) buffer).getBuf();
+          if (!closed) {
+            buf.retain();
+            results.add(buf);
+          }
+        }
+      }
+
+      @Override
+      public void onFailure(int chunkIndex, Throwable e) {
+        String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
+        logger.error(errorMsg, e);
+        exception.set(new IOException(errorMsg, e));
+      }
+    };
+    client = new RetryingChunkClient(conf, shuffleKey, location,
+      callback, clientFactory, startMapIndex, endMapIndex);
+    numChunks = client.openChunks();
+  }
+
+  public boolean hasData() {
+    return returnedChunks < numChunks && numChunks >= 1;
```

Comment:
No need for numChunks >=1. returnedChunks default to 0, If num Chunks is 0, returnedChunks < numChunks will be false